### DISCUSSION TITLE
[dashboard]Monitor Query on SuperSet Dashboard

### DISCUSSION THREAD

**Discussion by Asturias-sam** [apache/superset#18431] [Ideas]
Hello Team,

Is there any way to get the execution time of the queries on dashboard, so it will be helpful for performance analysis.
From SQLLAB we can get the execution time but for dashboards is it possible see the execution time for queries ?

**Comment by junlincc** [comment]
@Asturias-sam could you provide a use case? thanks

**Comment by Asturias-sam** [comment]
@junlincc 
We have dashboard having various charts, we want to monitor Superset performance  i.e  latency between Superset and Our data sources in terms of query execution shouldn't be much.

**Comment by junlincc** [comment]
If your goal is to improve dashboard performance, I suggest that to checkout and enable async query first, docs can be found [here ](https://superset.apache.org/docs/installation/async-queries-celery)  If you feel like testing the feature, feel free to enable the GLOBAL_ASYNC_QUERIES feature flag. See the CONTRIBUTING.md documentation for details: https://github.com/apache/superset/blob/master/CONTRIBUTING.md#async-chart-queries

I'm still trying to understand the business value of this requested feature. I assume you wanna use it as an assessment indicator against other BI tools?

**Comment by Asturias-sam** [comment]
We want to manage the Dashboard Performance  in Superset so it will help us to alert our team about slow performance and we  can take the necessary steps to speed up our datasources.

**Comment by joaopmedeiros** [comment]
Im trying to the same thing and include errors..

**Comment by yitian-reevo** [comment]
Hi folks, callback from 2025, is there any updates for this topic?

My team is expecting the same thing, we build our reporting page based on SuperSet with 10+ charts and we would like to monitor the performance of each chart, e.g. the request time of superset api, the sql execution time for each query etc.

**Comment by YuriyKrasilnikov** [comment]
## Query Lifecycle Timing Breakdown for Chart Data Requests

### Problem

When a dashboard chart loads slowly, it is currently impossible to determine where the time is spent within Superset. The database query log may show a fast query (e.g. 200ms), yet the chart takes 10+ seconds to load. The overhead could be in Jinja template rendering, security/RLS checks, query compilation, result serialization, or connection pool wait time.

Without phase-level breakdown, debugging slow dashboards is guesswork. In containerized deployments, this can cause cascading failures — a Celery worker becomes unresponsive during a long-running phase, liveness probes fail, the container restarts, and the root cause remains invisible.

### Industry Comparison

| Tool | Total Time | Phase Breakdown | UI Access |
|------|-----------|----------------|----------|
| [Looker](https://cloud.google.com/looker/docs/query-performance-metrics) | Yes | Yes (3 phases: Init / Query / Processing) | [Performance Panel](https://cloud.google.com/looker/docs/query-tracker) |
| [Metabase](https://www.metabase.com/docs/latest/usage-and-performance-tools/usage-analytics) | Yes | No | Usage Analytics (paid) |
| Redash | Yes | No | No |
| **Superset** | Partial (SQL Lab only) | No | No |

### Proposed Approach

Superset already has the infrastructure: `STATS_LOGGER` with `timing()`, `stats_timing` context manager, `@statsd_metrics` decorator on API endpoints (total time only).

Proposal: wrap each phase of the `/api/v1/chart/data` query lifecycle with `stats_timing` (~6-8 instrumentation points):

```
query_context_processor.py:get_df_payload()
  ├─ cache lookup            → stats_timing("chart_data.cache_lookup")
  ├─ jinja template render   → stats_timing("chart_data.jinja_rendering")
  ├─ security/RLS checks     → stats_timing("chart_data.security_checks")
  ├─ database execution      → stats_timing("chart_data.db_execution")
  └─ result serialization    → stats_timing("chart_data.result_processing")
```

Include a `timing` object in the chart data API response:

```json
{
  "result": [...],
  "timing": {
    "cache_lookup_ms": 5,
    "jinja_rendering_ms": 120,
    "security_checks_ms": 45,
    "db_execution_ms": 850,
    "result_processing_ms": 200,
    "total_ms": 1220
  }
}
```

Configurable slow query logging when total exceeds a threshold. Minimal overhead — timestamp calls only. All existing infrastructure (`STATS_LOGGER`, `stats_timing`) is reused.

Related: Discussion #13044
